### PR TITLE
Jiterator with Python Registration

### PR DESCRIPTION
**Pull request by SherlockNoMad**
You can now do a lot of crazy things about redefining the behavior of an operator, and still be fast in cuda !!! 

Example 1: swapping where's branches
```
code_string = "template <typename T> T inverted_where(bool cond, T a, T b){ return !cond ? a : b; }"
jitted_fn = torch.cuda.jiterator._create_jit_fn(code_string) 
my_lib = torch.library.Library("aten", "IMPL")
my_lib.impl('aten::where.self', jitted_fn, "CUDA")

# torch.where is now overridden
```
Example 2: approximate gelu with relu
```
code_string = "template <typename T> T fast_gelu(T a){ return a > 0 ? a : 0;}"
jitted_fn = torch.cuda.jiterator._create_jit_fn(code_string) 
my_lib = torch.library.Library("aten", "IMPL")
my_lib.impl('aten::gelu', jitted_fn, "CUDA")

# torch.nn.GELU and torch.nn.function.gelu are now overridden
```
Example 3: clipping output for numerical unstable kernels 
```
code_string = "template <typename T> T clipped_exp(T a){ return a > T(10.0) ? T(22026.4657948) : exp(a); }"
jitted_fn = torch.cuda.jiterator._create_jit_fn(code_string) 
my_lib = torch.library.Library("aten", "IMPL")
my_lib.impl('aten::exp', jitted_fn, "CUDA")

# torch.exp(x) and x.exp() are now overridden
```
Example 4: Simulate buggy hardware behaviors 
```
code_string = "template <typename T> T buggy_add(T a, T b){ return a + b + T(1); }"
jitted_fn = torch.cuda.jiterator._create_jit_fn(code_string) 
my_lib = torch.library.Library("aten", "IMPL")
my_lib.impl('aten::add.Tensor', jitted_fn, "CUDA")

torch.add(x, y), "x + y" and x.add(y) are now overridden
```

### CODE REVIEW

File: test/test_python_dispatch.py [hunk 1]
```diff
@@ -79,6 +80,111 @@ def my_sum(*args, **kwargs):
...
+            '''
+            jitted_add = _create_jit_fn(buggy_add_code_string)
+
+            CALLED = [False]
+            def buggy_add(*args, **kwargs):
+                CALLED[0] = True
+                return jitted_add(*args, **kwargs)
+
+            my_lib = Library("aten", "IMPL")
+            my_lib.impl('aten::add.Tensor', buggy_add, "CUDA")
```
Comment (ezyang) [inline test/test_python_dispatch.py:170]: You need to manually ensure the implementation gets deregistered or it will be permanently broken. cc @anjali411

Comment (ezyang) [reply test/test_python_dispatch.py:170]: Given the current design of operator registration, del does not work, you have to use remove method

Comment (SherlockNoMad) [reply test/test_python_dispatch.py:170]: I see in the __ del__ function of Library, it removes all the registered impls automatically. 
@anjali411, could you please help me understand if anything is still missing with the "del my_lib"

File: test/test_python_dispatch.py [hunk 2]
```diff
@@ -79,6 +80,117 @@ def my_sum(*args, **kwargs):
...
+            my_lib.impl('aten::where.self', inverted_where, "CUDA")
+
+            device = 'cuda'
+            cond = torch.tensor([True, True, False], device=device, dtype=torch.bool)
+            x = torch.tensor([1, 2, 3], device=device)
+            y = torch.tensor([-1, -2, -3], device=device)
+
+            self.assertEqual(torch.where(cond, x, y), torch.tensor([-1, -2, 3]))
+            self.assertTrue(CALLED[0])
+            del my_lib
```
Comment (anjali411) [inline test/test_python_dispatch.py:110]: I guess it's nice to have explicit deletes but they are not necessary

Comment (SherlockNoMad) [reply test/test_python_dispatch.py:110]: yep.

Comment (anjali411) [reply test/test_python_dispatch.py:110]: add a check that the old behavior is retained after your library is deleted

File: test/test_python_dispatch.py [hunk 3]
```diff
@@ -79,6 +80,117 @@ def my_sum(*args, **kwargs):
...
+
+            x = torch.rand([3, 3], device='cpu')
+            y = torch.rand([3], device='cpu')
+
+            x_cuda = x.cuda()
+            y_cuda = y.cuda()
+
+            self.assertEqual(x_cuda + y_cuda, x + y + 1)
+            self.assertTrue(CALLED[0])
+            del my_lib
```
Comment (anjali411) [inline test/test_python_dispatch.py:186]: add a check that the old behavior is retained after your library is deleted

**Review by anjali411 — APPROVED** [review]

File: test/test_python_dispatch.py (hunk 3 above)
Comment (SherlockNoMad) [reply test/test_python_dispatch.py:195]: done.

**Pull request closed without merging**